docs/architecture/adr-007-address-generation-of-chain-link-proof.md

```diff
@@ -0,0 +1,157 @@
+# ADR 007: Address generation of chain link proof
+
+## Changelog
+
+- January 20th, 2022: Initial draft;
+- July 28th, 2022: First review;
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+Currently, Desmos allows linking other chains accounts which addresses are formatted using either the `Bech32`, `Base58` or `Hex` encoding and generated using a single algorithm specific to the encoding itself. Since Desmos idea is to support as many chains as possible, we SHOULD split the address generation algorithm from the encoding algorithm so that more chains can be linked properly.
+
+## Context
+
+The `x/profiles` module gives users the possibility to link their profile to different external accounts.  However, each one of the supported address formats currently supports only one generation algorithm.
+
+`Bech32Address` relies on the address generation algorithm specific of Cosmos-SDK chains:
+- for single signature accounts: 
+  ```
+  ripemd160(sha256(32 bytes public key))[:20]
+  ``` 
+- for multi-sig accounts: 
+  ```
+  sha256(aminoCdc.Marshal(multisig public key))[:20]
+  ```
+
+`Base58Address` relies on the Solana algorithm:
+```
+(32 bytes public key)[:32]
+```
+
+`HexAddress` relies on the Ethereum algorithm:
+```
+keccak64(64 bytes public key)[12:]
+```
+
+Due to the fact all the `AddressData` implementations are highly coupled to a single address generation function, it is currently impossible to link an address that needs to be encoded using a supported encoding method but is generated using a different algorithm. This is the case of Elrond addresses which are encoded using the `Bech32` encoding algorithm, but are generated using an algorithm that is different from the Cosmos one.
+
+## Decision
+
+In order to allow developers to integrate any kind of address, we will review how the current `Address` structure is made. Instead of using an interface with multiple implementations based on the encoding algorithm, we will use a single structure that allows to specify both the encoding algorithm and the hashing algorithm(s) that should be used to get the proper address: 
+
+```protobuf
+syntax = "proto3";
+
+// Address contains the data of an external address
+message Address {
+  // Hex-encoded value of the address
```

Comment:
Instead of requesting the developers to put the hex-encoded address bytes here, we can also allow them to put the already-encoded address using the specified encoding algorithm. Then, inside the `Validate` check we simply: 
1. generate the addresss bytes from the public key; 
2. encode the address using the provided encoding method; 
3. check if the resulting encoded address is equals to the provided value